server/persistence/mongodb/mongodb.go

```diff
@@ -0,0 +1,127 @@
+package mongodb
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/bsoncodec"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+
+	"github.com/joshprzybyszewski/cribbage/model"
+	"github.com/joshprzybyszewski/cribbage/server/interaction"
+	"github.com/joshprzybyszewski/cribbage/server/persistence"
+	"github.com/joshprzybyszewski/cribbage/server/persistence/mongodb/mapbson"
+)
+
+const (
+	dbName     string = `cribbage`
+	gamesCol   string = `games`
+	playersCol string = `players`
+)
+
+var _ persistence.DB = (*mongodb)(nil)
+
+type mongodb struct {
+	bsonRegistry *bsoncodec.Registry
+	client       *mongo.Client
+}
+
+func New(uri string) (persistence.DB, error) {
+	if uri == `` {
+		// If we don't know where to connect, use the default localhost URI
+		uri = `mongodb://localhost:27017`
+	}
+	ctx, _ := context.WithTimeout(context.Background(), 10*time.Second)
+	client, err := mongo.Connect(ctx, options.Client().ApplyURI(uri))
+	if err != nil {
+		return nil, err
+	}
+
+	return &mongodb{
+		client:       client,
+		bsonRegistry: mapbson.CustomRegistry(),
+	}, nil
+}
+
+func (m *mongodb) gamesCollection() *mongo.Collection {
+	gColOpts := []*options.CollectionOptions{{
+		Registry: m.bsonRegistry,
+	}}
+	return m.client.Database(dbName).Collection(gamesCol, gColOpts...)
+}
+
+func (m *mongodb) playersCollection() *mongo.Collection {
+	gColOpts := []*options.CollectionOptions{{
+		Registry: m.bsonRegistry,
+	}}
+	return m.client.Database(dbName).Collection(playersCol, gColOpts...)
+}
+
+func (m *mongodb) GetGame(id model.GameID) (model.Game, error) {
+	result := model.Game{}
+	ctx, _ := context.WithTimeout(context.Background(), 5*time.Second)
+	err := m.gamesCollection().FindOne(ctx, bson.M{"id": id}).Decode(&result)
+
+	if err != nil {
+		return model.Game{}, err
+	}
+
+	return result, nil
+}
+
+func (m *mongodb) GetPlayer(id model.PlayerID) (model.Player, error) {
+	result := model.Player{}
+	ctx, _ := context.WithTimeout(context.Background(), 5*time.Second)
+	err := m.playersCollection().FindOne(ctx, bson.M{"id": id}).Decode(&result)
+
+	if err != nil {
+		return model.Player{}, err
+	}
+	return result, nil
+}
+
+func (m *mongodb) GetInteraction(id model.PlayerID) (interaction.Player, error) {
+	return nil, errors.New(`unimplemented`)
+}
+
+func (m *mongodb) SaveGame(g model.Game) error {
+	ctx, _ := context.WithTimeout(context.Background(), 5*time.Second)
+	// TODO insert this like the memory does, where we append it to the end of a list of games,
```

Comment:
We need to append the new game state to the saved list of previous game states...

We do this so that we can incrementally build Actions onto a game. I'm starting to wonder whether the `Action` field on `model.Game` should be public...